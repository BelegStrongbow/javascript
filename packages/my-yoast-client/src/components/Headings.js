import styled from "styled-components";
import colors from "yoast-components/style-guide/colors.json";
import defaults from "../config/defaults.json";

export const ListHeading = styled.span`
	display: block;
	font-size: 1em;
	padding: 24px 40px;
	font-weight: 400;
	margin: 0;
	border-bottom: 1px solid ${ colors.$color_grey_medium };

	@media screen and ( max-width: ${ defaults.css.breakpoint.tablet }px ) {
		padding: 24px;
	}

	@media screen and ( max-width: ${ defaults.css.breakpoint.mobile }px ) {
		padding: 16px;
	}
`;

export const Heading = styled.h1`
	font-weight: 700;
	font-size: 2em;
`;

export const ModalHeading = styled.h1`
	font-weight: 300;
	font-size: 1.5em;
	margin: 0;
`;

export const SubHeading = styled.h2`
	font-weight: 300;
	font-size: 1.3em;
	margin: 0;
`;
