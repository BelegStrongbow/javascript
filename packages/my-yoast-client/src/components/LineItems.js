import PropTypes from "prop-types";
import React from "react";
import styled from "styled-components";
import defaults from "../config/defaults.json";

const ItemsContainer = styled.ul`
	padding: 0;
	list-style: none;
	flex-shrink: 0;

	@media screen and ( max-width: ${ defaults.css.breakpoint.mobile }px ) {
		overflow: hidden;
	}
`;

/**
 * LineItems.
 *
 * @param {Object} props Properties to be passed to the table.
 * @returns {ReactElement} A row of LineItems.
 */
function LineItems( props ) {
	const items = props.items.map( ( item ) => {
		return <li key={ item.id }>{ item.quantity }<span aria-hidden="true">&times;</span> { item.productName }</li>;
	} );
	return <ItemsContainer>{ items }</ItemsContainer>;
}

LineItems.propTypes = {
	items: PropTypes.array,
};

LineItems.defaultProps = {
	items: [],
};

export default LineItems;
