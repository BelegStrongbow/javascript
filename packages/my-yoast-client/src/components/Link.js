import PropTypes from "prop-types";
import React, { Component } from "react";
import { Link as RouterLink } from "react-router-dom";

import NewTabMessage from "./NewTabMessage";

/**
 * Determines if a certain URL is an external URL.
 *
 * @param {string} url The URL to test.
 * @returns {boolean} Whether or not the URL is an external URL.
 */
function isExternal( url ) {
	return url.startsWith( "http" ) || url === "#" || url.startsWith( "mailto:" );
}

/**
 * Determines if a certain URL points to Yoast.
 *
 * @param {string} url The URL to test.
 * @returns {boolean} Whether or not the URL points to Yoast.
 */
function isYoastLink( url ) {
	return /yoast\.com|yoast\.test|yoa\.st/.test( url );
}

/**
 * Link that also works for external URL's, see https://github.com/ReactTraining/react-router/issues/1147
 * for more information.
 *
 * When an external link has a `target="_blank"` attribute, it will automatically
 * add the accessibility visually hidden message "(Opens in a new browser tab)"
 * by using the `NewTabMessage` component.
 *
 */
export default class Link extends Component {
	/**
	 * Renders the component.
	 *
	 * @returns {ReactElement} The rendered component.
	 */
	render() {
		const internalProps = Object.assign( {}, this.props );
		delete internalProps.linkTarget;
		delete internalProps.linkRel;
		delete internalProps.ariaLabel;
		delete internalProps.enabledStyle;
		delete internalProps.iconSource;
		delete internalProps.iconSize;

		/*
		 * When a link has a target _blank attribute and it doesn't point to Yoast,
		 * we always want a default "noopener" rel attribute value.
		 */
		let relValueForTargetBlank = "noopener";

		/*
		 * When a link has a target _blank attribute and it does point to Yoast,
		 * we use the linkRel prop which defaults to `null` so it doesn't render
		 * the rel attribute. This way, it can be overridden setting the prop, if necessary.
		 */
		if ( isYoastLink( this.props.to ) ) {
			relValueForTargetBlank = this.props.linkRel;
		}

		const hasTargetBlank = this.props.linkTarget === "_blank";

		const externalProps = {
			href: this.props.to,
			className: this.props.className,
			"aria-label": this.props.ariaLabel,
			target: this.props.linkTarget,
			rel: hasTargetBlank ? relValueForTargetBlank : this.props.linkRel,
			onClick: this.props.onClick,
		};

		return isExternal( this.props.to )
			? <a { ...externalProps }>
				{ this.props.children }
				{ hasTargetBlank && <NewTabMessage /> }
			</a>
			: <RouterLink { ...internalProps } />;
	}
}

Link.propTypes = {
	to: PropTypes.string.isRequired,
	className: PropTypes.string,
	children: PropTypes.any.isRequired,
	ariaLabel: PropTypes.string,
	linkTarget: PropTypes.string,
	linkRel: PropTypes.string,
	onClick: PropTypes.func,
};

Link.defaultProps = {
	className: null,
	ariaLabel: null,
	linkTarget: null,
	linkRel: null,
	onClick: null,
};
