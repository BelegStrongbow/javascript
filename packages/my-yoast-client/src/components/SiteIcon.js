import styled from "styled-components";

const SiteIcon = styled.img`
	height: inherit;
`;

export default SiteIcon;
