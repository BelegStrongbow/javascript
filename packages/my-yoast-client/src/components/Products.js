import PropTypes from "prop-types";
import React from "react";
import styled from "styled-components";
import colors from "yoast-components/style-guide/colors.json";
import Product from "./Product";
import { ListHeading } from "./Headings";

const ProductsContainer = styled.div`
	display: flex;
	background-color: ${ colors.$color_white };
	padding: 0 24px 24px 24px;
	margin: 0 0 0 8px;
	flex-wrap: wrap;
	width: 100%;
	align-items: stretch;
`;

const ProductsListHeading = styled( ListHeading )`
	width: calc( 100% - 16px );
	font-size: 22px;
	margin: 0 0 24px 0px;
	padding: 0 0 24px 0;
`;

/**
 * Returns the rendered Products component.
 *
 * @param {Object} props The props to use.
 *
 * @returns {ReactElement} The rendered products page.
 */
export default class Products extends React.Component {
	/**
	 * Renders the component.
	 *
	 * @returns {ReactElement} The rendered component.
	 */
	render() {
		const props = this.props;
		return (
			<ProductsContainer>
				<ProductsListHeading>
					{ props.heading }
					{ props.byLine }
				</ProductsListHeading>
				{ props.noResults }
				{ props.products.map( function( product ) {
					return <Product
						key={ product.glNumber }
						name={ product.name }
						glNumber={ product.glNumber }
						currentVersion={ product.currentVersion }
						icon={ product.icon }
						buttons={ product.buttons }
						composerToken={ props.composerToken }
						onComposerHelpModalOpen={ props.onComposerHelpModalOpen }
						onComposerHelpModalClose={ props.onComposerHelpModalClose }
						type={ product.category }
					/>;
				} ) }
			</ProductsContainer>
		);
	}
}

Products.propTypes = {
	byLine: PropTypes.element,
	heading: PropTypes.string.isRequired,
	noResults: PropTypes.string,
	composerToken: PropTypes.object,
	onComposerHelpModalOpen: PropTypes.func,
	onComposerHelpModalClose: PropTypes.func,
	composerHelpProductName: PropTypes.string,
};

Products.defaultProps = {
	byLine: null,
	noResults: "",
	onComposerHelpModalClose: () => {},
	onComposerHelpModalOpen: () => {},
};
