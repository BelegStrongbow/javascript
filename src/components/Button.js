import React, { Component } from "react";
import styled from "styled-components";
import colors from "yoast-components/style-guide/colors.json";
import Link from "./Link";
import angleLeft from "../icons/angle-left.svg";
import chevronRight from "../icons/chevron-right.svg";
import questionCircle from "../icons/question-circle.svg";
import defaults from "../config/defaults.json";

let buttonAnimations = `
	transition: background 150ms ease-out;

	&:focus,
	&:hover {
		box-shadow: 0 4px 10px 0 rgba(0, 0, 0, 0.2), inset 0 0 0 100px rgba( 0, 0, 0, 0.1 );
	}

	&:active {
		transform: translateY( 1px );
		box-shadow: none;
	}
`;

export const Button = styled.button`
	height: 48px;
	margin: 0;
	// Buttons don't need vertical padding.
	padding: 0 16px;
	border: 0;
	background-color: ${ props => props.enabledStyle ? colors.$color_green_medium_light : colors.$color_grey_disabled };
	color: ${ colors.$color_white };
	box-shadow: 0 2px 8px 0 rgba(0, 0, 0, 0.3);
	border-radius: 4px;
	font: 400 14px/24px "Open Sans", sans-serif;
	text-transform: uppercase;
	cursor: pointer;

	${ buttonAnimations };
`;

Button.propTypes = {
	onClick: React.PropTypes.func,
	type: React.PropTypes.string,
	enabledStyle: React.PropTypes.bool,
};

Button.defaultProps = {
	type: "button",
	enabledStyle: true,
};

export const LargeButton = styled( Button )`
	min-width: 152px;
`;

export const LargeSecondaryButton = styled( LargeButton )`
	color: ${ colors.$color_black };
	box-shadow: none;
	background-color: ${ colors.$color_white };
	border: 1px solid ${ colors.$color_black };
	
	&:focus,
	&:hover {
		background-color: ${ colors.$color_green_medium_light };
		color: ${ colors.$color_white };
		box-shadow: 0 2px 8px 0 rgba(0, 0, 0, 0.3);
		border: none;
	}
`;

export const TextButton = styled( Button )`
	width: ${ props => props.buttonWidth };
`;

TextButton.PropTypes = {
	buttonWidth: React.PropTypes.string,
	enabledStyle: React.PropTypes.bool,
};

TextButton.defaultProps = {
	buttonWidth: "auto",
	enabledStyle: true,
};

export const LogoutButton = styled( Button )`
	border-radius: 3px;
	height: 36px;
	width: 112px;
	padding: 0;
`;

export const IconButton = styled( Button )`
	background-repeat: no-repeat;
	background-image: url( ${ props => props.iconSource } );
	background-position: 16px 50%;
	background-size: ${ props => props.iconSize };
	padding-left: 56px;
`;

IconButton.PropTypes = {
	iconSource: React.PropTypes.string.isRequired,
	iconSize: React.PropTypes.string,
	enabledStyle: React.PropTypes.bool,
};

IconButton.defaultProps = {
	iconSize: "24px",
	enabledStyle: true,
};

export const MobileHeaderButton = styled( IconButton )`
	background-color: transparent;
	box-shadow: none;
	display: block;
	position: fixed;

	&:hover {
		box-shadow: none;
	}

	@media screen and ( max-width: ${ defaults.css.breakpoint.mobile }px ) {
		text-indent: -9999em;
		box-shadow: none;
	}
`;

export const LargeIconButton = styled( IconButton )`
	min-width: 152px;
`;

export const IconButtonTransparent = styled( IconButton )`
	background-color: transparent;
	background-position: 0.5em 50%;
	color: ${ colors.$color_blue };
	box-shadow: none;
	text-transform: none;
	height: 32px;
	padding-left: 2em;

	transition: background 150ms ease-out;

	&:focus,
	&:hover {
		box-shadow: none;
		background-color: ${ colors.$color_grey_light };
	}

	&:active {
		transform: translateY( 1px );
		box-shadow: none;
	}
`;

export const WhiteButton = styled( LargeButton )`
	color: ${ colors.$color_blue };
	background-color: ${ colors.$color_white };
`;

export const RedButton = styled( LargeButton )`
	background-color: ${ colors.$color_red };
`;

export const ButtonLink = styled( Link )`
	display: inline-block;
	height: 48px;
	padding: 12px 16px;
	background-color: ${ colors.$color_green_medium_light };
	color: ${ colors.$color_white };
	box-shadow: 0 2px 8px 0 rgba(0, 0, 0, 0.3);
	border-radius: 4px;
	// line height 24 + padding 12 + 12 = 48
	font: 400 14px/24px "Open Sans", sans-serif;
	text-transform: uppercase;
	text-decoration: none;
	text-align: center;

	${ buttonAnimations }
`;

ButtonLink.PropTypes = {
	to: React.PropTypes.string.isRequired,
};

export const LargeButtonLink = styled( ButtonLink )`
	min-width: 152px;
`;

export const IconButtonLink = styled( ButtonLink )`
	background-repeat: no-repeat;
	background-image: url( ${ props => props.iconSource } );
	background-position: 8px 50%;
	background-size: ${ props => props.iconSize };
	padding-left: 36px;
`;

IconButtonLink.PropTypes = {
	iconSource: React.PropTypes.string.isRequired,
	iconSize: React.PropTypes.string,
	enabledStyle: React.PropTypes.bool,
};

IconButtonLink.defaultProps = {
	iconSize: "24px",
	enabledStyle: true,
};

export const LargeIconButtonLink = styled( IconButtonLink )`
	min-width: 152px;
`;

export const BackButtonLink = styled( LargeIconButtonLink )`
	background-image: url( ${ angleLeft } );
`;

export const TextButtonLink = styled( ButtonLink )`
	width: ${ props => props.buttonWidth };
`;

TextButtonLink.PropTypes = {
	buttonWidth: React.PropTypes.string,
};

TextButtonLink.defaultProps = {
	buttonWidth: "auto",
};

export const ChevronButton = styled( Button )`
	background-color: transparent;
	background-repeat: no-repeat;
	background-image: url( ${ chevronRight } );
	background-position: center;
	background-size: 32px;
	width:  48px;
	height: 48px;
	cursor: pointer;
	box-shadow: none;
`;

ChevronButton.propTypes = {
	onClick: React.PropTypes.func,
	type: React.PropTypes.string,
	"aria-label": React.PropTypes.string,
};

ChevronButton.defaultProps = {
	type: "button",
	"aria-label": "",
};

export const BeaconButton = styled( Button )`
	border-radius: 0;
	height: 36px;
	width: 152px;
	font-size: 16px;
	cursor: pointer;
	position: fixed;
	z-index: 1;
	left: -112px;
	top: 85%;
	color: ${ colors.$color_white };
	background-color: ${ colors.$color_green_medium_light };
	z-index: 1;
	padding-right: 40px;
	background-repeat: no-repeat;
	background-image: url( ${ questionCircle } );
	background-position: right 8px center;
	background-size: 24px;
	transition: left .5s ease-in;
	text-shadow: 0 0 2px rgba(0, 0, 0, 0.6);

	&:focus,
	&:hover {
		left: 0;
		box-shadow: 0 4px 10px 0 rgba(0, 0, 0, 0.2);
	}

	@media screen and ( max-width: 1024px ) {
		display: none;
	}
`;

BeaconButton.propTypes = {
	onClick: React.PropTypes.func,
	type: React.PropTypes.string,
	"aria-label": React.PropTypes.string,
};

BeaconButton.defaultProps = {
	type: "button",
	"aria-label": "",
};

export const IconRightButtonLink = styled( ButtonLink )`
	background-repeat: no-repeat;
	background-image: url( ${ props => props.iconSource } );
	background-position: right 16px center;
	background-size: 24px;
	// 8px grid: 16 left background position + 24 icon size + 16 = 56
	padding-left: 16px;
	padding-right: 56px;
`;

IconRightButtonLink.PropTypes = {
	iconSource: React.PropTypes.string.isRequired,
};

export const WhiteButtonLink = styled( LargeButtonLink )`
	background-color: ${ colors.$color_white };
	color: ${ colors.$color_blue };
`;

/**
 * Returns a disabled button component for the given component.
 *
 * @param {React.Component} Button The button.
 * @returns {React.Component} The disabled button.
 */
export function disable( Button ) {
	let StyledDisabledButton = styled( Button )`
		background-color: ${ colors.$color_grey_disabled };
		cursor: default;

		&:focus,
		&:hover {
			box-shadow: none;
		}

		&:active {
			transform: none;
		}
	`;

	return class DisabledButton extends Component {
		render() {
			return <StyledDisabledButton disabled="disabled" { ...this.props } />;
		}
	};
}

/**
 * Makes a button full width in the mobile responsive view.
 *
 * @param {ReactElement} component The original button.
 * @returns {ReactElement} The button with full width responsive style.
 */
export function makeButtonFullWidth( component ) {
	return styled( component )`
		@media screen and ( max-width: ${ defaults.css.breakpoint.mobile }px ) {
			width: 100%;
			background-image: none;
			padding: 12px 16px;
		}
	`;
}

/**
 * Makes an icon button display only the icon in the intermdiate responsive view.
 *
 * @param {ReactElement} component The original button.
 * @returns {ReactElement} The button with icon only.
 */
export function makeResponsiveIconButton( component ) {
	return styled( component )`
		.screen-reader-text {
			position: static;
		}

		@media screen and (min-width: ${ defaults.css.breakpoint.mobile + 1 }px) and (max-width: ${ defaults.css.breakpoint.tablet }px) {
			padding-right: 0;
			padding-left: 42px;

			.screen-reader-text {
				position: absolute;
			}
		}
	`;
}
