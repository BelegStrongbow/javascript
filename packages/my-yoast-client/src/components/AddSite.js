import PropTypes from "prop-types";
import React from "react";
import { injectIntl, intlShape, defineMessages, FormattedMessage } from "react-intl";
import { LargeButton, makeButtonFullWidth, LargeSecondaryButton } from "./Button.js";
import styled from "styled-components";
import colors from "yoast-components/style-guide/colors.json";
import { addPlaceholderStyles } from "../styles/inputs";
import validate from "validate.js";
import { speak } from "@wordpress/a11y";
import _debounce from "lodash/debounce";
import ErrorDisplay from "../errors/ErrorDisplay";
import { ModalHeading } from "./Headings";
import ButtonsContainer from "./general/ButtonsContainer";
import YoastNativeSelect from "./general/YoastNativeSelect";
import { StyledLabel } from "./Labels";

const messages = defineMessages( {
	validationFormatURL: {
		id: "validation.format.url",
		defaultMessage: "Please enter a valid URL. Remember to use http:// or https://.",
	},
} );

const AddSiteModal = styled.div`
	max-width: 640px;
	margin: auto;
	font-size: 1em;
`;

const WebsiteURL = addPlaceholderStyles( styled.input`
	width: 100%;
	height: 48px;
	box-shadow: inset 0 2px 8px 0px rgba(0,0,0,0.3);
	background: ${ colors.$color_grey };
	padding: 0 0 0 10px;
	font-size: 1em;
	border: 0;
	margin-bottom: 8px;
` );

const ValidationText = styled.div`
	font-size: 1em;
	color: ${ colors.$color_red};
	margin: 1em 0;
	min-height: 1.8em;
`;

const WideLargeButton = makeButtonFullWidth( LargeButton );
const WideSecondaryButton = makeButtonFullWidth( LargeSecondaryButton );

/**
 * Returns the AddSite component.
 *
 * @returns {ReactElement} The AddSite component.
 */
class AddSite extends React.Component {
	/**
	 * Initializes the class with the specified props.
	 *
	 * @param {Object} props The props to be passed to the class that was extended from.
	 *
	 * @returns {void}
	 */
	constructor( props ) {
		super( props );

		this.constraints = {
			url: this.urlConstraints.bind( this ),
		};

		this.state = {
			validationError: null,
			showValidationError: false,
			selectedValue: "wordpress",
		};

		// Defines the debounced function to show the validation error.
		this.showValidationMessageDebounced  = _debounce( this.showValidationMessage, 1000 );
		// Defines the debounced function to announce the validation error.
		this.speakValidationMessageDebounced = _debounce( this.speakValidationMessage, 1000 );

		this.onWebsiteURLChange = this.onWebsiteURLChange.bind( this );
		this.handleOnSubmit     = this.handleOnSubmit.bind( this );
		this.handleOnBlur       = this.handleOnBlur.bind( this );
	}

	/**
	 * Tests whether customer has access to composer token feature. If so, returns a drop-down selector for CMS type.
	 * If not, returns null.
	 *
	 * @returns {*} Either null or html for the CMS type drop-down.
	 */
	getPlatformSelect() {
		return (
			<div>
				<StyledLabel
					htmlFor="add-site-select-platform"
				>
					<FormattedMessage
						id="sites.addSite.enterUrl"
						defaultMessage="Please select the platform that your website is running on:"
					/>
				</StyledLabel>
				<YoastNativeSelect
					selectId="add-site-select-platform"
					selectName="selectPlatform"
					selectDefaultValue={ this.state.selectedValue }
					selectOnBlur={ this.handleOnBlur }
				>
					<option value="wordpress">WordPress</option>
					<option value="typo3">TYPO3</option>
				</YoastNativeSelect>
			</div>
		);
	}

	/**
	 * Sets the constraints for validation to URL, and sets the message that should be returned if the constraints are not met.
	 *
	 * @returns {Object} Returns the constraints for the URL, and the message.
	 */
	urlConstraints() {
		return {
			url: {
				message: this.props.intl.formatMessage( messages.validationFormatURL ),
			},
		};
	}

	/**
	 * Calls onChange function when website url changes.
	 *
	 * @param {Object} event The event returned by the WebsiteURLChange.
	 *
	 * @returns {void}
	 */
	onWebsiteURLChange( event ) {
		const value = event.target.value;
		this.props.onChange( value );
	}

	/**
	 * Runs url validation and shows/hides error if validation returns error.
	 *
	 * @param {string}  url       The url to validate.
	 * @param {boolean} debounced Whether to show the debounced error message.
	 *
	 * @returns {void}
	 */
	runValidation( url, debounced = true ) {
		const validationError = this.validateUrl( url );
		if ( validationError ) {
			this.updateValidationMessage( validationError );
			if ( debounced ) {
				this.showValidationMessageDebounced();
				this.speakValidationMessageDebounced();
			} else {
				this.showValidationMessage();
				this.speakValidationMessage();
			}
		} else {
			this.updateValidationMessage( null );
			this.hideValidationError();
		}
	}

	/**
	 * Shows the validation error.
	 *
	 * @returns {void}
	 */
	showValidationMessage() {
		this.setState( { showValidationError: true } );
	}

	/**
	 * Updates the validation message.
	 *
	 * @param {string} message The validation message.
	 *
	 * @returns {void}
	 */
	updateValidationMessage( message ) {
		this.setState( {
			validationError: message,
		} );
	}

	/**
	 * Validates URL and shows validation error if URL is invalid.
	 *
	 * @param {string} input The URL to be validated.
	 * @returns {string} URL Validation error message.
	 */
	validateUrl( input = "" ) {
		if ( input === "" ) {
			return null;
		}

		const result = validate( { website: input }, { website: this.urlConstraints() }, { format: "detailed" } );

		if ( result && result[ 0 ] !== null ) {
			return result[ 0 ].options.message;
		}
		return null;
	}

	/**
	 * Hides the validation message and cancels a potential debounced error.
	 *
	 * @returns {void}
	 */
	hideValidationError() {
		this.setState( { showValidationError: false }, () => {
			this.showValidationMessageDebounced.cancel();
			this.speakValidationMessageDebounced.cancel();
		} );
	}

	/**
	 * Renders the validation error message container.
	 *
	 * When there are no errors, this container is rendered empty and acts like
	 * a "placeholder" taking space in the page to avoid "jumps" in the layout.
	 *
	 * @param {string} input The string in the input field.
	 * @returns {ReactElement} Returns a div that is either empty or contains an error message.
	 */
	urlValidityMessage( input = "" ) {
		if ( this.state.showValidationError && input !== "" ) {
			return (
				<ValidationText>
					<FormattedMessage
						id="sites.addSite.urlValidationMessage"
						defaultMessage={ this.state.validationError }
					/>
				</ValidationText>
			);
		}
		return (
			<ValidationText />
		);
	}

	/**
	 * Handles the submit event.
	 *
	 * @param {object} event The submit event.
	 *
	 * @returns {void}
	 */
	handleOnSubmit( event ) {
		event.preventDefault();
		if ( ! this.state.validationError && !! this.props.linkingSiteUrl ) {
			this.props.onConnectClick( this.state.selectedValue );
		}
	}

	/**
	 * Handles the blur event.
	 *
	 * @param {object} event The blur event.
	 *
	 * @returns {void}
	 */
	handleOnBlur( event ) {
		this.setState( {
			selectedValue: event.target.value,
		} );
	}

	/**
	 * Sends a message to the ARIA live assertive region.
	 *
	 * @returns {void}
	 */
	speakValidationMessage() {
		const message = this.props.intl.formatMessage( messages.validationFormatURL );
		speak( message, "assertive" );
	}

	/**
	 * Populates the URL input field with the search term, if any, and runs validation.
	 *
	 * @returns {void}
	 */
	componentDidMount() {
		if ( this.props.query ) {
			this.props.onChange( this.props.query );
			this.runValidation( this.props.query, false );
		}
	}

	/**
	 * Validates the URL when entering a value in the URL input field.
	 *
	 * @param {object} nextProps The new props the component will receive.
	 *
	 * @returns {void}
	 */
	componentWillReceiveProps( nextProps ) {
		if ( this.props.linkingSiteUrl !== nextProps.linkingSiteUrl ) {
			this.runValidation( nextProps.linkingSiteUrl );
		}
	}

	/**
	 * Cancels debounced functions when the component unmounts.
	 *
	 * @returns {void}
	 */
	componentWillUnmount() {
		this.showValidationMessageDebounced.cancel();
		this.speakValidationMessageDebounced.cancel();
	}

	/**
	 * Renders the component.
	 *
	 * @returns {ReactElement} The rendered component.
	 */
	render() {
		return (
			<AddSiteModal>
				<ModalHeading>
					<FormattedMessage id="sites.addSite.header" defaultMessage="Add Site" />
				</ModalHeading>

				<form onSubmit={ this.handleOnSubmit } noValidate={ true }>
					<StyledLabel htmlFor="add-site-input">
						<FormattedMessage
							id="sites.addSite.enterUrl"
							defaultMessage="Please enter the URL of the site you would like to add to your account:"
						/>
					</StyledLabel>

					<WebsiteURL
						type="url"
						id="add-site-input"
						placeholder={ "https://example-site.com" }
						value={ this.props.linkingSiteUrl }
						onChange={ this.onWebsiteURLChange }
					/>

					<ErrorDisplay error={ this.props.error } />

					{ this.getPlatformSelect() }
					{ this.urlValidityMessage( this.props.linkingSiteUrl ) }
					<ButtonsContainer>
						<WideSecondaryButton onClick={ this.props.onCancelClick }>
							<FormattedMessage id="sites.addSite.cancel" defaultMessage="Cancel" />
						</WideSecondaryButton>
						<WideLargeButton
							type="submit"
							enabledStyle={ ! this.state.validationError && !! this.props.linkingSiteUrl }
						>
							<FormattedMessage id="sites.addSite.connect" defaultMessage="Add" />
						</WideLargeButton>
					</ButtonsContainer>
				</form>
			</AddSiteModal>
		);
	}
}

AddSite.propTypes = {
	intl: intlShape.isRequired,
	linkingSiteUrl: PropTypes.string.isRequired,
	onCancelClick: PropTypes.func.isRequired,
	onConnectClick: PropTypes.func.isRequired,
	onChange: PropTypes.func.isRequired,
	query: PropTypes.string.isRequired,
	error: PropTypes.object,
};

AddSite.defaultProps = {
	error: {},
};

export default injectIntl( AddSite );
