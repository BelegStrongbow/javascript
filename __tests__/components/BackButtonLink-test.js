
import React from 'react';
import renderer from 'react-test-renderer';
import { BrowserRouter as Router } from 'react-router-dom';
import { BackButtonLink } from '../../src/components/Button';

test('the round back button matches the snapshot', () => {
	const component = renderer.create(
		<Router>
			<BackButtonLink to="/" >ButtonValue</BackButtonLink>
		</Router>
	);

	let tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});

test('the round back button with button type given matches the snapshot', () => {
	const component = renderer.create(
		<Router>
			<BackButtonLink to="/" type="button">ButtonValue</BackButtonLink>
		</Router>
	);

	let tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});
