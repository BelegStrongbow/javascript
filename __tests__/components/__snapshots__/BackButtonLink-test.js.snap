// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the round back button matches the snapshot 1`] = `
<a
  className="jyTnCM"
  href="/"
  onClick={[Function]}
>
  ButtonValue
</a>
`;

exports[`the round back button with button type given matches the snapshot 1`] = `
<a
  className="jyTnCM"
  href="/"
  onClick={[Function]}
  type="button"
>
  ButtonValue
</a>
`;
